src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+    -- timestamp should be integer
+    local int_ts = nsec ~= nil or usec ~= nil or msec ~= nil
+
+    local ts = obj.timestamp
+    local fraction
+    if ts ~= nil then
+        s, fraction = math_modf(ts)
+        if not int_ts then
+            nsec = fraction * 1e9
+        end
+        hms = true
+    end
+
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, {-720, 720}, offset)
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        y = y or 1970
+        M = M or 1
+        if d ~= nil and d < 0 then
+            d = builtin.tnt_dt_days_in_month(y, M)
+        end
+        dt = builtin.tnt_dt_from_ymd(y, M, d)
+    end
+
+    -- .hour, .minute, .second
+    local secs = 0
+    if hms then
+        secs = (h or 0) * 3600 + (m or 0) * 60 + (s or 0)
+    end
+
+    return datetime_new_dt(dt, secs, nsec, offset or 0)
+end
+
+local sz = 48
+local buff = ffi.new('char[?]', sz)
+
+--[[
+    Convert to text datetime values
+
+    - datetime will use ISO-8601 format:
+        1970-01-01T00:00Z
+        2021-08-18T16:57:08.981725+03:00
+]]
+local function datetime_tostring(self)
+    check_date(self, 'datetime.tostring()')
+
+    local len = builtin.datetime_to_string(buff, sz, self)
+    assert(len < sz)
+    return ffi.string(buff)
+end
+
+--[[
+    Basic    Extended
+    Z        N/A
+    +hh      N/A
+    -hh      N/A
+    +hhmm    +hh:mm
+    -hhmm    -hh:mm
+
+    Returns pair of constructed datetime object, and length of string
+    which has been accepted by parser.
+]]
+parse_zone = function(str)
+    check_str("datetime.parse_zone()")
+    local offset = ffi.new('int[1]')
+    local len = builtin.tnt_dt_parse_iso_zone_lenient(str, #str, offset)
+    if len == 0 then
+        error(('invalid time-zone format %s'):format(str), 3)
+    end
+    return offset[0]
+end
+
+--[[
+    Dispatch function to create datetime from string or table.
+    Creates default timeobject (pointing to Epoch date) if
+    called without arguments.
+]]
+local function datetime_from(o)
+    if o == nil or type(o) == 'table' then
+        return datetime_new(o)
+    end
+end
+
+--[[
+    Create datetime object representing current time using microseconds
+    platform timer and local timezone information.
+]]
+local function local_now()
+    local d = datetime_new_raw(0, 0, 0)
+    builtin.datetime_now(d)
+    return d
+end
+
+--[[
+    Return table in os.date('*t') format, but with timezone
+    and nanoseconds
+]]
+local function datetime_totable(self)
+    local secs = local_secs(self) -- hour:minute should be in local timezone
+    local dt = local_dt(self)
+
+    return {
+        year = builtin.tnt_dt_year(dt),
+        month = builtin.tnt_dt_month(dt),
+        yday = builtin.tnt_dt_doy(dt),
+        day = builtin.tnt_dt_dom(dt),
+        wday = (ffi.cast('int32_t', builtin.tnt_dt_dow(dt)) + 1) % 7,
```

Comment:
Why do you need this cast?

Reply:
Because `tnt_dt_dow()` returns `dt_dow_t` which is enumeration, and we need a cast (AFAIK) to get the normal underlying integer value:
```
tarantool> dt = 719163
---
...

tarantool> ffi = require 'ffi'
---
...

tarantool> ffi.C.tnt_dt_dow(dt)
---
- 'cdata<enum 967>: 4'
...

tarantool> ffi.cast('int32_t', ffi.C.tnt_dt_dow(dt))
---
- 4
...
```